pkg/comp-functions/functions/vshnnextcloud/collabora.go

```diff
@@ -0,0 +1,597 @@
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      comp.GetName() + "-collabora-code-issuer",
+			Namespace: comp.GetInstanceNamespace(),
+		},
+		Spec: cmv1.IssuerSpec{
+			IssuerConfig: cmv1.IssuerConfig{
+				SelfSigned: &cmv1.SelfSignedIssuer{},
+			},
+		},
+	}
+
+	return svc.SetDesiredKubeObject(issuer, comp.GetName()+"-collabora-code-issuer", runtime.KubeOptionAddLabels(labelMap))
+}
+
+func createCertificate(comp *vshnv1.VSHNNextcloud, svc *runtime.ServiceRuntime) error {
+	certificate := &cmv1.Certificate{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      comp.GetName() + "-collabora-code-certificate",
+			Namespace: comp.GetInstanceNamespace(),
+		},
+		Spec: cmv1.CertificateSpec{
+			SecretName: comp.GetName() + "-collabora-code-tls",
+			DNSNames:   []string{comp.Spec.Parameters.Service.Collabora.FQDN},
+			IssuerRef: certmgrv1.ObjectReference{
+				Name: comp.GetName() + "-collabora-code-issuer",
+			},
+		},
+	}
+
+	return svc.SetDesiredKubeObject(certificate, comp.GetName()+"-collabora-code-certificate", runtime.KubeOptionAddLabels(labelMap))
+}
+
+func createCoolWSDConfigMap(comp *vshnv1.VSHNNextcloud, svc *runtime.ServiceRuntime) error {
+	cm := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      comp.GetName() + "-collabora-code-coolwsd-config",
+			Namespace: comp.GetInstanceNamespace(),
+		},
+		Data: map[string]string{
+			"coolwsd.xml": coolwsd,
+			"coolkit.xml": coolkit,
+		},
+	}
+
+	return svc.SetDesiredKubeObject(cm, comp.GetName()+"-collabora-code-coolwsd-config", runtime.KubeOptionAddLabels(labelMap))
+}
+
+/*
+Generate RSA keys for Collabora Code
+necessary for full WOPI support, please refer to: https://learn.microsoft.com/en-us/microsoft-365/cloud-storage-partner-program/rest/
+*/
+func createSecretWithRSAKeys(comp *vshnv1.VSHNNextcloud, svc *runtime.ServiceRuntime) error {
+	objName := comp.GetName() + "-collabora-code-rsa-keypair"
+
+	currentSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      objName,
+			Namespace: comp.GetInstanceNamespace(),
+		},
+	}
+
+	err := svc.GetObservedKubeObject(currentSecret, objName)
+	if err == runtime.ErrNotFound {
+		/*
+		 this part might look weird, but it's necessary because:
+		 - we don't want to overwrite the secret if it already exists
+		 - RSA generation is really slow
+		*/
+		return svc.SetDesiredKubeObject(currentSecret, objName, runtime.KubeOptionAddLabels(labelMap))
+	} else if err != nil {
+		return err
+	}
+
+	privkey, pubkey, err := generateSSHRSAKey()
+	if err != nil {
+		return err
+	}
+
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      objName,
+			Namespace: comp.GetInstanceNamespace(),
+		},
+		Data: map[string][]byte{
+			"id_rsa":     []byte(privkey),
+			"id_rsa.pub": []byte(pubkey),
+		},
+	}
+
+	return svc.SetDesiredKubeObject(secret, objName, runtime.KubeOptionAddLabels(labelMap))
+}
+
+func createWatchOnlyObject(comp *vshnv1.VSHNNextcloud, svc *runtime.ServiceRuntime) error {
```

Comment:
I would change the name of this function to smth like, `observeCertManagedCertificate()` or whatever secret you are observing instead of `createWatchOnlyObject()`. It's more clear what the function will do and maybe the comment might not be necessary to read.

Reply:
I changed that function name as well